src/mmap.rs

```diff
@@ -45,15 +51,42 @@ pub(crate) trait AsSlice {
 
 /// Errors that can happen when creating a memory map
 #[derive(Debug)]
-pub enum MmapError {
-    /// Syscall returned the given error.
-    SystemCallFailed(io::Error),
+pub enum Error {
+    /// Adding the guest base address to the length of the underlying mapping resulted
+    /// in an overflow.
+    InvalidGuestRegion,
+    /// Error creating a `MmapRegion` object.
+    MmapRegion(MmapRegionError),
     /// No memory region found.
     NoMemoryRegion,
     /// Some of the memory regions intersect with each other.
     MemoryRegionOverlap,
 }
 
+// TODO: use this for Windows as well after we redefine the Error type there.
+#[cfg(unix)]
+/// For a borrowed `FileOffset` and size, this function checks whether the mapping does not
+/// extend past EOF, and that adding the size to the file offset does not lead to overflow.
+pub fn check_file_offset(
+    file_offset: &FileOffset,
+    size: usize,
+) -> result::Result<(), MmapRegionError> {
+    let file = file_offset.file();
+    let start = file_offset.start();
+
+    if let Some(end) = start.checked_add(size as u64) {
+        if let Ok(metadata) = file.metadata() {
+            if metadata.len() < end {
+                return Err(MmapRegionError::MappingPastEof);
+            }
+        }
```

Comment:
Do we need to handle the case when file.metadata() fails?

Reply:
We could create a special Error case for it.
Once @alexandruag updates it, we need to re-approve the PR.

Reply:
We could merge the big PR first and submit another bugfix PR for this issue.
Review a small bugfix PR is much easier:)